nano/node/repcrawler.cpp

```diff
@@ -1,370 +1,524 @@
+				}
+			},
+			nano::transport::buffer_drop_policy::no_socket_drop);
+		}
+		else
+		{
+			logger.debug (nano::log::type::rep_crawler, "Ignoring duplicate query for block {} to {}", hash_root.first.to_string (), channel->to_string ());
+			stats.inc (nano::stat::type::rep_crawler, nano::stat::detail::query_duplicate);
 		}
 	}
-	return error;
 }
 
-nano::uint128_t nano::rep_crawler::total_weight () const
+void nano::rep_crawler::query (std::shared_ptr<nano::transport::channel> const & target_channel)
 {
-	nano::lock_guard<nano::mutex> lock{ probable_reps_mutex };
-	nano::uint128_t result (0);
-	for (auto i (probable_reps.get<tag_account> ().begin ()), n (probable_reps.get<tag_account> ().end ()); i != n; ++i)
+	query (std::vector{ target_channel });
+}
+
+bool nano::rep_crawler::is_pr (std::shared_ptr<nano::transport::channel> const & channel) const
+{
+	nano::lock_guard<nano::mutex> lock{ mutex };
+	auto existing = reps.get<tag_channel> ().find (channel);
+	if (existing != reps.get<tag_channel> ().end ())
 	{
-		if (i->channel->alive ())
-		{
-			result += node.ledger.weight (i->account);
-		}
+		return node.ledger.weight (existing->account) >= node.minimum_principal_weight ();
 	}
-	return result;
+	return false;
 }
 
-void nano::rep_crawler::on_rep_request (std::shared_ptr<nano::transport::channel> const & channel_a)
+bool nano::rep_crawler::process (std::shared_ptr<nano::vote> const & vote, std::shared_ptr<nano::transport::channel> const & channel)
 {
-	nano::lock_guard<nano::mutex> lock{ probable_reps_mutex };
-	if (channel_a->get_tcp_endpoint ().address () != boost::asio::ip::address_v6::any ())
+	nano::lock_guard<nano::mutex> lock{ mutex };
+
+	auto & index = queries.get<tag_channel> ();
+	auto [begin, end] = index.equal_range (channel);
+	for (auto it = begin; it != end; ++it)
 	{
-		probably_rep_t::index<tag_channel_ref>::type & channel_ref_index = probable_reps.get<tag_channel_ref> ();
+		// TODO: This linear search could be slow, especially with large votes.
+		auto const target_hash = it->hash;
+		bool found = std::any_of (vote->hashes.begin (), vote->hashes.end (), [&target_hash] (nano::block_hash const & hash) {
+			return hash == target_hash;
+		});
 
-		// Find and update the timestamp on all reps available on the endpoint (a single host may have multiple reps)
-		auto itr_pair = channel_ref_index.equal_range (*channel_a);
-		for (; itr_pair.first != itr_pair.second; itr_pair.first++)
+		if (found)
 		{
-			channel_ref_index.modify (itr_pair.first, [] (nano::representative & value_a) {
-				value_a.last_request = std::chrono::steady_clock::now ();
-			});
+			logger.debug (nano::log::type::rep_crawler, "Processing response for block {} from {}", target_hash.to_string (), channel->to_string ());
+			stats.inc (nano::stat::type::rep_crawler, nano::stat::detail::response);
+			// TODO: Track query response time
+
+			responses.push_back ({ channel, vote });
+			queries.erase (it);
```

Comment:
The query cannot be erased here; if a node has multiple PRs served by one channel then votes after the first will not be processed by the rep crawler due to the query being deleted by the first vote.

Reply:
Yeah, I have a pretty good idea on how to fix this. In the meantime it would be great if you could add your testcase to the branch.

Reply:
We should not be actively deleting the query because we cannot possibly know when to delete it.
We should just let it timeout.

Reply:
Yes, this is an easy and obvious fix.